src/main/java/com/amazon/opendistroforelasticsearch/security/dlic/rest/api/AbstractApiAction.java

```diff
@@ -606,10 +594,23 @@ protected boolean isValidRolesMapping(final RestChannel channel, final String ro
 		}
 
 		if (isReadOnly(rolesMappingConfiguration, role)) {
-			forbidden(channel, "Role '" + role + "' has read-only role-mapping.");
+			conflict(channel, "Role '" + role + "' has read-only role-mapping.");
```

Comment:
I don't think that `409` is the correct status code here. Note that  according to spec `Conflict` "is used in situations where the user might be able to resolve the conflict and resubmit the request" and `Forbidden` "might be forbidden for reasons unrelated to the credentials."

Reply:
As per [wiki](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_client_errors) and resources 405 is recommended for PUT on a read-only resource so will change to that.

Reply:
Please check ES behavior. It uses 403 for read-only indices. It will be better to keep the current status code as both 403 and 405 are acceptable.

Reply:
I see elasticsearch returning 403 for AuthorizationException but 405/409 seems to commonly used for read-only resources so its okay to change to 405

Reply:
User has authorization to perform the request but the resource is read-only. Returning 405 would differentiate the behavior

Reply:
I refer to ES behavior when `index.blocks.write` is set to `true`:
```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "cluster_block_exception",
        "reason" : "index [readonly] blocked by: [FORBIDDEN/8/index write (api)];"
      }
    ],
    "type" : "cluster_block_exception",
    "reason" : "index [readonly] blocked by: [FORBIDDEN/8/index write (api)];"
  },
  "status" : 403
}
```

Reply:
Slightly torn on this. Why is it necessary to follow ES while most common practice is to return 405 ?

Reply:
AFAIK, there is no common practice and both 403 and 405 are used. Security plugin is part of ES deployment, so being consistent with ES is more important (and also avoids breaking REST API backward compatibility with prior security plugin versions).

Reply:
Is returning 403 for readonly resource an ES practice only? Don't see it being commonly used

Reply:
Returning 403 for a read only resource when user has correct authn & authz to perform the request feels wrong hence still preferring to return 405

Reply:
> Is returning 403 for readonly resource an ES practice only? Don't see it being commonly used

No, it is not.

> Returning 403 for a read only resource when user has correct authn & authz to perform the request feels wrong hence still preferring to return 405

Specification explicitly defines that 403 may be returned "for reasons unrelated to the credentials."

I don't see a need to change the behavior of the API from 403 to 405 unless there is a specific ask from a security plugin users.

Reply:
Returning appropriate status code will differentiate behavior when user has correct authorization but modifying read only resource and there I see a need. 
I shall open up an issue for this and based on user discussion we can make/retain the change.

Reply:
https://github.com/opendistro-for-elasticsearch/security/issues/719